common/flatpak-run.c

```diff
@@ -3165,6 +3165,29 @@ flatpak_run_app (const char     *app_ref,
       g_snprintf (pid_str, sizeof (pid_str), "%d", child_pid);
       pid_path = g_build_filename (instance_id_host_dir, "pid", NULL);
       g_file_set_contents (pid_path, pid_str, -1, NULL);
+
+      g_clear_pointer (&bwrap, flatpak_bwrap_free); // this drops our lock
+
+      {
+        g_autofree char *ref_file = g_build_filename (instance_id_host_dir, ".ref", NULL);
+
+        struct flock l = {
+          .l_type = F_WRLCK,
+          .l_whence = SEEK_SET,
+          .l_start = 0,
+          .l_len = 0
+        };
+        glnx_autofd int lock_fd = open (ref_file, O_RDWR | O_CLOEXEC);
+        if (lock_fd != -1)
+          {
+            while (fcntl (lock_fd, F_GETLK, &l) == 0 && l.l_type == F_UNLCK)
+              {
+                g_debug ("Waiting for bwrap to take the lock");
+                g_usleep (10000);
+                l.l_type = F_WRLCK;
+              }
```

Comment:
If anything bad happens in the bwrap this would loop forever though. We probably want *some* kind of timeout.

Reply:
right